### PR TITLE
GSoC'22 Enzyme New pass manager 

### LINKED ISSUE #584: Migrate to new Pass Manager
https://llvm.org/docs/NewPassManager.html

The tricky part is to keep our custom command line options working.

### PR DESCRIPTION
**Pull request by ki9gpin**
**Project Update** 
Date : September 12th , 2022
The new PassManager is now functional in Enzyme besides the legacy PassManager .  So enzyme can now be used with both the pass managers. Tests (.ll files) remain to be updated so that they would also test Enzyme for the new pass manager.

**---------------------------------------------------------**
Date : April 17th , 2022
I am  applying for GSoC'22 Enzyme New pass manager integration project. This pull request is to present my candidacy. Any suggestion would be greatly helpful.

fixes #584

### CODE REVIEW

File: enzyme/BCLoad/BCLoader.cpp [hunk 1]
```diff
@@ -6,6 +6,9 @@
 #include "llvm/Support/SourceMgr.h"
 #include "llvm/Support/raw_ostream.h"
 
+#include "BCLoader.h"
+#include "BCLoaderNew.h"
```
Comment (wsmoses) [inline enzyme/BCLoad/BCLoader.cpp:10]: Can you put these in a single header?

File: enzyme/BCLoad/ClangBCLoader.cpp [hunk 2]
```diff
@@ -3,6 +3,8 @@
 
 #include "BCLoader.h"
 
+#include "BCLoaderNew.h"
+
```
Comment (wsmoses) [inline enzyme/BCLoad/ClangBCLoader.cpp:7]: We will also need to register the new pass manager version to run automatically like the old pass manager version below.

File: enzyme/BCLoad/BCLoader.cpp [hunk 3]
```diff
@@ -14,13 +17,8 @@ using namespace llvm;
...
 
-namespace {
-class BCLoader : public ModulePass {
-public:
-  static char ID;
-  BCLoader() : ModulePass(ID) {}
 
-  bool runOnModule(Module &M) override {
+bool runCommon(Module& M){
     std::set<std::string> bcfuncs = {"cblas_ddot"};
```
Comment (wsmoses) [inline enzyme/BCLoad/BCLoader.cpp:22]: Can the name of this function be changed to something more descriptive, like say “lowerDeclarationsToBitcode”?

File: enzyme/BCLoad/BCLoader.cpp (hunk 1 above)
Comment (ki9gpin) [reply enzyme/BCLoad/BCLoader.cpp:10]: Should I create another header or  move the new code into the old header?

Comment (wsmoses) [reply enzyme/BCLoad/BCLoader.cpp:10]: Into the old header is fine

File: enzyme/BCLoad/BCLoader.cpp (hunk 3 above)
Comment (ki9gpin) [reply enzyme/BCLoad/BCLoader.cpp:22]: yeah , sure.

File: enzyme/BCLoad/BCLoader.cpp (hunk 1 above)
Comment (ki9gpin) [reply enzyme/BCLoad/BCLoader.cpp:10]: ok

File: enzyme/BCLoad/ClangBCLoader.cpp (hunk 2 above)
Comment (ki9gpin) [reply enzyme/BCLoad/ClangBCLoader.cpp:7]: Need help . How do I register a module pass . Couldn't find in documentation about ModulePass. Then searched for in llvm code.

Comment (wsmoses) [reply enzyme/BCLoad/ClangBCLoader.cpp:7]: https://medium.com/@mshockwave/writing-pass-instrument-for-llvm-newpm-f17c57d3369f

This blog should have some relevant information

Comment (ki9gpin) [reply enzyme/BCLoad/ClangBCLoader.cpp:7]: Sorry. Load shedding is going  on , stormy night, I am using  my phone . 
I will update these asap.

Comment (ki9gpin) [reply enzyme/BCLoad/ClangBCLoader.cpp:7]: > https://medium.com/@mshockwave/writing-pass-instrument-for-llvm-newpm-f17c57d3369f
> 
> This blog should have some relevant information

Thanks William.

Comment (ki9gpin) [reply enzyme/BCLoad/ClangBCLoader.cpp]: downgraded to llvm-11 . Now check-bcpass test pass. here is SS of /cblas_ddot_double.bc file not found error
Screenshot_20220418_070659

Comment (wsmoses) [reply enzyme/BCLoad/ClangBCLoader.cpp]: I think you need to pass -bcpath=/path/to/bclib

Comment (ki9gpin) [reply enzyme/BCLoad/ClangBCLoader.cpp]: Should I start writing the proposal already?

Comment (wsmoses) [reply enzyme/BCLoad/ClangBCLoader.cpp]: Yes, most definitely

Comment (ki9gpin) [reply enzyme/BCLoad/ClangBCLoader.cpp]: > I think you need to pass -bcpath=/path/to/bclib

 'bcpath'  is an unknown argument to opt.

Comment (ki9gpin) [reply enzyme/BCLoad/ClangBCLoader.cpp]: > Yes, most definitely
20th april is deadline for proposal it's already 19 .
Will you be able to review my proposal before  submission and give me a feedback?

Comment (ki9gpin) [reply enzyme/BCLoad/ClangBCLoader.cpp]: manually setting the path to the bitcode file successfully loads the bitcode
Screenshot_20220419_181427

**Comment by ki9gpin** [comment]
Hi @wsmoses @vchuravy 
According to 
[this](https://github.com/EnzymeAD/Enzyme/blob/12ebbf48ea6fa7156bef9095232fd747dbf40f23/enzyme/BCLoad/BCLoader.cpp#L84)
line  analysis is not preserved ,right? Thus I should return PreservedAnalysis::none() for the new pass manager ?

I noticed 
cmake build was not  building subdirectory BCLoad  .
**ENZYME_EXTERNAL_SHARED_LIB**  cmake option flag  is set to **OFF**  and changing the  value to  **ON**  cmake also builds BCLoad subdirectory .

**Comment by ki9gpin** [comment]
Facing this issue
image

File: enzyme/BCLoad/BCLoader.cpp [hunk 4]
```diff
@@ -14,13 +20,8 @@ using namespace llvm;
 cl::opt<std::string> BCPath("bcpath", cl::init(""), cl::Hidden,
```
[the remark this answers is not in the document]
Comment (ki9gpin) [reply enzyme/BCLoad/BCLoader.cpp]: command line options working for "BCPass" which uses new PM . But can't invoke "bcloader" which uses legacy PM .
Here's an image 
image

**Comment by ki9gpin** [comment]
It's working . Produces same result. Checked for  ( .bc) and also for ( .ll) files as input.

For New PM
opt-12 -load ./BCLoad/BCPass-12.so  -load-pass-plugin=./BCLoad/BCPass-12.so ../bclib/cblas_ddot_double.bc -S -passes=BCPass   > output-cblas-ddot-npm.ll

For Legacy PM
opt-12 -load ./BCLoad/BCPass-12.so  -load-pass-plugin=./BCLoad/BCPass-12.so ../bclib/cblas_ddot_double.bc -S  -bcloader  > output-cblas-ddot.ll

Diff returns empty
diff output-cblas-ddot.ll output-cblas-ddot-npm.ll

File: enzyme/BCLoad/BCLoader.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (ki9gpin) [reply enzyme/BCLoad/BCLoader.cpp]: It's working . Produces same result. Checked for ( .bc) and also for ( .ll) files as input.

For New PM
opt-12 -load ./BCLoad/BCPass-12.so -load-pass-plugin=./BCLoad/BCPass-12.so ../bclib/cblas_ddot_double.bc -S -passes=BCPass > output-cblas-ddot-npm.ll

For Legacy PM
opt-12 -load ./BCLoad/BCPass-12.so -load-pass-plugin=./BCLoad/BCPass-12.so ../bclib/cblas_ddot_double.bc -S -bcloader > output-cblas-ddot.ll

Diff returns empty
diff output-cblas-ddot.ll output-cblas-ddot-npm.ll

[the remark this answers is not in the document]
Comment (ki9gpin) [reply enzyme/BCLoad/BCLoader.cpp]: image

File: .devcontainer/devcontainer.json [hunk 5]
```diff
@@ -9,5 +9,5 @@
         "source=enzyme-extensions-insiders,target=/root/.vscode-server-insiders/extensions,type=volume",
         "source=enzyme-build,target=${containerWorkspaceFolder}/enzyme/build,type=volume",
     ],
-    "postCreateCommand": "sudo chown vscode ./enzyme/build"
+    "postCreateCommand": "sudo git config --global --add safe.directory /workspaces/Enzyme &&  sudo chown vscode ./enzyme/build"
```
[the remark this answers is not in the document]
Comment (ki9gpin) [reply .devcontainer/devcontainer.json:12]: Git was not functional inside devcontainer. Added this line and  git was functional again.

File: .gitignore [hunk 6]
```diff
@@ -3,3 +3,4 @@
 *.out
 build
 .vscode
+CLArgs.txt
```
[the remark this answers is not in the document]
Comment (ki9gpin) [reply .gitignore]: Removed the line

**Comment by wsmoses** [comment]
@ki9gpin can you update the test to also use the new PM, then lets merge this

File: enzyme/test/BCLoader/fblas_ddot.ll [hunk 7]
```diff
@@ -1,5 +1,6 @@
 ;RUN: if [ %llvmver -ge 10 && %llvmver -le 12 ]; then %clang %s -Xclang -load -Xclang %loadBC -S -emit-llvm -o - | %FileCheck %s; fi
 ;RUN: if [ %llvmver -ge 12 ]; then %clang %s -fno-experimental-new-pass-manager -Xclang -load -Xclang %loadBC -S -emit-llvm -o - | %FileCheck %s; fi
+;RUN: if [ %llvmver -ge 12 ]; then %opt %s -load-pass-plugin=%loadBC -passes="bcloader" -S | %FileCheck %s; fi
```
Comment (wsmoses) [inline enzyme/test/BCLoader/fblas_ddot.ll:3]: Can you have a version of this also use clang rather than opt?

**Comment by ki9gpin** [comment]
hi @wsmoses  
the pass registration method _registeroptimizerearlyepcallback_ is not available in llvm 12. What  I found is, it was added in llvm 15. Would you take a look .

File: .devcontainer/devcontainer.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (tgymnich) [reply .devcontainer/devcontainer.json]: The usual workflow would be to make commits outside the dev container.

[the remark this answers is not in the document]
Comment (ki9gpin) [reply .devcontainer/devcontainer.json]: As i'm not using devcontainer now , this issue can hang in there for a while. Changing back.

**Pull request closed without merging**